docs/datamodel/build-a-datamodel.rst

```diff
@@ -0,0 +1,626 @@
+                    },
+                }
+            }
+        }
+    }
+
+The above Elasticsearch mapping, similarly to the JSONSchema, defines the
+structure of the JSON, but also how it should be indexed.
+
+For instance in above example the ``title`` field is of type ``text`` which
+applies stemming when searching, whereas the ``keywords`` field is of type
+``keyword`` which means no stemming is applied. The mapping also allows you to
+define e.g. that a ``lat`` and a ``lon`` field are in fact geographical
+coordinates, and enable geospatial queries over your records.
+
+Naming JSONSchemas and mappings
+-------------------------------
+You may already have noticed that both JSONSchemas and Elasticsearch mappings
+are using the same folder structure and naming scheme:
+
+.. code-block:: shell
+
+    |-- my_datamodel
+    |   |-- jsonschemas
+    |   |   |-- __init__.py
+    |   |   `-- records
+    |   |       `-- record-v1.0.0.json
+    |   |-- mappings
+    |   |   |-- __init__.py
+    |   |   `-- v6
+    |   |       |-- __init__.py
+    |   |       `-- records
+    |   |           `-- record-v1.0.0.json
+
+
+The naming scheme is very important for two reasons:
+
+1. Indexing of records
+2. Data model evolution
+
+**Indexing of records**
+
+Invenio will determine the Elasticsearch index for a given record, based on the
+record's ``$schema`` key. For instance, given the following record:
+
+.. code-block:: json
+
+    {
+        "$schema": "https://localhost/schemas/records/record-v1.0.0.json",
+        "...": "..."
+    }
+
+Invenio will send the above record to the ``records-record-v1.0.0``
+Elasticsearch index. Note, it's possible to customize this behavior.
+
+**Data model evolution**
+
+Over time data models are likely to evolve. In many cases, you can simply make
+backward compatible changes to the existing JSONSchema and Elasticsearch
+mappings. In cases, where you change the data model in a backward incompatible
+way, you create a new JSONSchema and new mappings (e.g. ``record-v1.1.0.json``)
+
+.. code-block:: shell
+
+    |-- my_datamodel
+    |   |-- jsonschemas
+    |   |   |-- __init__.py
+    |   |   `-- records
+    |   |       `-- record-v1.0.0.json
+    |   |       `-- record-v1.1.0.json
+    |   |-- mappings
+    |   |   |-- __init__.py
+    |   |   `-- v6
+    |   |       |-- __init__.py
+    |   |       `-- records
+    |   |           `-- record-v1.0.0.json
+    |   |           `-- record-v1.1.0.json
+
+
+This allows you to simultaneously store old an new records - i.e. you don't
+have take down your service for hours to migrate you millions of records from
+one version to a new.
+
+Now of course, old records will be sent to the ``records-record-v1.0.0`` index
+and new records will be sent to the ``records-record-v1.1.0`` index. However,
+a special Elasticsearch *index alias* ``records`` is also created, that allows
+you to search over both old an new records, thus smoothly handling data model
```

Comment:
old an(d) new

Reply:
Typo fixes up to here: https://github.com/dinosk/invenio/commit/a842f7011dbbfd252d6d64670426542fedff525b

Reply:
Thanks a lot